lib/xenops_server.ml

```diff
@@ -1328,15 +1331,40 @@ let rec immediate_operation dbg id op =
 			Xenops_task.run task
 		) ();
 	match Xenops_task.get_state task with
-		| Task.Pending _ -> assert false
-		| Task.Completed _ -> ()
-		| Task.Failed e ->
-			let e = e |> Exception.exnty_of_rpc |> exn_of_exnty in
-			raise e
+	| Task.Pending _ -> assert false
+	| Task.Completed _ -> ()
+	| Task.Failed e ->
+		let e = e |> Exception.exnty_of_rpc |> exn_of_exnty in
+		raise e
+
+and queue_operation_int ~progress_callback dbg id op =
+	let task = Xenops_task.add tasks dbg (let r = ref None in fun t -> perform ~result:r op t; !r) in
+	Redirector.push Redirector.parallel_queues id (op, task);
+	task
+
+and queue_operations_and_wait ~progress_callback ~max_parallel_atoms dbg id ops =
```

Comment:
Also, there are functions `queue_operation` and `queue_operation_and_wait` around L1715. Are we duplicating logic now? In any case, I am getting a little confused because the names are all so similar.